src/lib/pickles_types/higher_kinded_poly.ml

```diff
@@ -0,0 +1,352 @@
+(* Caution: We do some low-level type manipulation to extract the value here.
+   This is safe because:
+   * the input types [_ t] are parameterized over [t1, ..., tn] which are
+     distinct opaque types which cannot unify
+   * the polymorphic types are parameterized over [(t1, ..., tn) t], which can
+     only unify when the underlying parameterised type is the same
+   * the underlying representation of OCaml values is homogeneous, which is
+     what allows for polymorphism in the wider language.
+
+   This can be used to approximate 'modular explicits' using packed modules
+   instead of substitutive module application, but it is generally more
+   powerful because it makes parameterised types first class outside of arrows
+   too (e.g. in heterogeneous lists).
+*)
+
+open Core_kernel
+open Poly_types
+
+(* Opaque types. *)
+module O = struct
+  type t1
+
+  type t2
+
+  type t3
+
+  type t4
+
+  type t5
+end
+
+module P1 = struct
+  type (_, _) t
```

Comment:
Maybe add a comment describing this type as being like an "apply" type? That helped me think about the rest of this code.